oshi-core/src/main/java/oshi/software/common/AbstractOperatingSystem.java

```diff
@@ -136,29 +144,14 @@ private int queryPlatformBitness() {
      *            A collection of all processes
      * @param parentPid
      *            The process ID whose children or descendants to return
-     * @param descendantPids
-     *            On input, an empty set, or a set of other children already
-     *            retrieved when using recursion. On output, the children of the
-     *            parent added to the set.
-     * @param recurse
+     * @param allDescendants
      *            If false, only gets immediate children of this process. If true,
      *            gets all descendants.
+     * @return Set of children or descendants of parentPid
      */
-    protected static void addChildrenToDescendantSet(Collection<OSProcess> allProcs, int parentPid,
-            Set<Integer> descendantPids, boolean recurse) {
-        // Collect this process's children
-        Set<Integer> childPids = allProcs.stream().filter(p -> p.getParentProcessID() == parentPid)
-                .map(OSProcess::getProcessID).collect(Collectors.toSet());
-        // Add to descendant set
-        descendantPids.addAll(childPids);
-        // Recurse
-        if (recurse) {
-            for (int pid : childPids) {
-                if (pid != parentPid && !descendantPids.contains(pid)) {
-                    addChildrenToDescendantSet(allProcs, pid, descendantPids, true);
-                }
-            }
-        }
+    protected static Set<Integer> getChildrenOrDescendants(Collection<OSProcess> allProcs, int parentPid, boolean allDescendants) {
+        Map<Integer, Integer> parentPidMap = allProcs.stream().collect(Collectors.toMap(OSProcess::getProcessID, p -> p.getParentProcessID()));
```

Comment:
I knew there was some sort of stream map thing. Thanks for reminding me what it was. I like this approach. Is it not possible to do `toMap(OSProcess::getProcessID, OSProcess::getParentProcessID)`?  Also, would be good to add a filter before entering in the map to verify parent start time is equal to or before child start time.

Reply:
Also, using this for the Solaris branch is actually probably more efficient (a single `ps` command line) than the existing multiple `pgrep` command lines (at the cost of the extra ps parsing for everything else).   Might be useful in the "descendants" case in Solaris to just call this method, keeping the `pgrep` option only for children.

Reply:
I think I've got all of the other fixes sorted out besides these two.  For the start time filtering, I'm thinking I just need to change it to `allProcs.stream().filter( p -> p.getStartTime() >= ???  ).collect(Collectors.toMap(OSProcess::getProcessID, OSProcess::getParentProcessID))` but I'm not sure how to get the parent's start time.

For the Solaris part, would this change be correct (leaving everything else the same)?

```       
if (allDescendants) {
        if (allDescendants) {
            Collection<OSProcess> allProcs = getProcessListFromPS("ps -eo s,pid,ppid,user,uid,group,gid,nlwp,pri,vsz,rss,etime,time,comm,args", -1);
            descendantPids = getChildrenOrDescendants(allProcs, Integer.parseInt(parentPid), allDescendants)
                    .stream().map(String::valueOf).collect(Collectors.toSet());
        }
} 
```

Reply:
RE: start time... things are always [easier to say than to do](https://xkcd.com/1425/), eh?

Follow along with my train of thought here: 
 - It seems we'd need to have already mapped the processes to their start times using a similar iteration but with start time instead of parent pid.  
 - But rather than iterating twice we could just iterate once here and create a map of pid to OSProcess
 - This would change the map parameter passed to `getChildrenOrDescendants` to `Map<Integer, OSProcess>`
 - This would require changing the map generation in `LinuxOperatingSystem.getParentPidsFromProcFiles()` and `WindowsOperatingSystem.getParentPidsFromSnapshot()` to include start time.  But this is not trivial on Linux, and isn't in the snapshot on Windows.  (And in Linux it'd be faster to compare up times.)

So, nevermind.  Just leave it for now, maybe I'll make that another improvement at a later time.

RE... Solaris.  You've got two nested if's asking the same thing, but ignoring that what you have works but duplicates code.  Just call `queryAllProcesses()`.  So I think it'd just be:
```
if (allDescendants) {
    descendantPids = getChildrenOrDescendants(queryAllProcesses(), Integer.parseInt(parentPid), true)
                    .stream().map(String::valueOf).collect(Collectors.toSet());
}
```
Except also, don't directly use `parseInt` that could throw an exception, use ParseUtil.parseIntOrDefault() with a negative pid default that will not match any children.

Also I recall some obscure fact comparing `String.valueOf()` vs. `Integer.toString()` and localization/number format differences and preferring one over the other. I'll have to look that up.

Reply:
Ok, either of those is identical. `String.valueOf()` simply calls `Integer.toString()` directly (and one would assume the JIT compiler would produce identical bits).   The obscure thing I was thinking of was using `%d` in `String.format()` which can produce odd results. Using `%s` invokes `Integer.toString()` and produces consistent output (e.g., 1234 not 1,234).

Reply:
Woops sorry I didn't realize I duplicated the if when I copy/pasted the code here!

Using `queryAllProcesses()` wasn't working because of the issue of calling a non-static method from a static method, so that's why I did it that way.  What else can I do to get around this?

Sounds good, I'll change the `parseInt` part. Thanks!

Reply:
Just remove the static modifier.  :)

Reply:
Or create a common static method that both call.  The only reason that is not static is it implements an abstract method in the superclass.